src/core/xlio.h

```diff
@@ -382,6 +384,225 @@ int xlio_socketxtreme_ref_buff(struct xlio_buff_t *buff);
  */
 int xlio_socketxtreme_free_buff(struct xlio_buff_t *buff);
 
+/*
+ * Storage API
+ */
+
+typedef uintptr_t xlio_group_t;
+typedef uintptr_t xlio_socket_t;
+typedef uint32_t xlio_key_t;
+
+struct xlio_buf {
+    uint64_t userdata;
+};
+
+/*
+ * XLIO initialization.
+ *
+ * xlio_init_ex() must be called before using any storage API. This is heavy operation.
+ * xlio_init_ex() is not thread-safe operation, however, subsequent serialized calls exit
+ * successfully without any action.
+ *
+ * If set, memory_cb() notifies about memory blocks which zerocopy RX buffers can point to.
+ * Current implementation allocates a single memory block and does it within xlio_init_ex() context.
+ */
+
+/*
+ * Memory callback.
+ *
+ * XLIO calls the callback each time XLIO allocates a memory region which can be used for RX
+ * buffers. User can use this information to prepare the memory for some logic in the future.
+ * Zerocopy RX interface provides pointers to such memory.
+ *
+ * Current XLIO implementation does a single allocation for buffers.
+ */
+typedef void (*xlio_memory_cb_t)(void *addr, size_t len, size_t hugepage_size);
+
+struct xlio_init_attr {
+    unsigned flags;
+    xlio_memory_cb_t memory_cb;
+};
+
+int xlio_init_ex(const struct xlio_init_attr *attr);
+
+/*
+ * Socket callbacks.
+ */
+
+enum {
+    /* TCP connection established. */
+    XLIO_SOCKET_EVENT_ESTABLISHED = 1,
+    /* Socket terminated and no further events are possible. */
+    XLIO_SOCKET_EVENT_TERMINATED,
+    /* Passive close. */
+    XLIO_SOCKET_EVENT_CLOSED,
+    /* An error occurred, see the error code value. */
+    XLIO_SOCKET_EVENT_ERROR,
+};
+
+/*
+ * Socket event callback.
+ *
+ * May be called from xlio_group_poll() context.
+ * In the callback context, send operation is allowed only for the ESTABLISHED event.
+ * Argument value holds the error code for the ERROR event and 0 for other events.
+ *
+ * List of possible error code values:
+ * ECONNABORTED - connection aborted by local side
+ * ECONNRESET - connection reset by remote side
+ * ECONNREFUSED - connection refused by remote side during TCP handshake
+ * ETIMEDOUT - connection timed out due to keepalive, user timeout option or TCP handshake timeout
+ */
+typedef void (*xlio_socket_event_cb_t)(xlio_socket_t, uintptr_t userdata_sq, int event, int value);
+
+/*
+ * Zerocopy completion event.
+ *
+ * May be called from the following contexts:
+ *  - xlio_group_poll() - likely
+ *  - xlio_socket_send() - can happen only if data is flushed
+ *  - xlio_socket_flush() / xlio_group_flush()
+ *  - xlio_socket_destroy()
+ *
+ * In the callback context, send operation is allowed unless the socket is under destruction.
+ */
+typedef void (*xlio_socket_comp_cb_t)(xlio_socket_t, uintptr_t userdata_sq, uintptr_t userdata_op);
```

Comment:
Should we provide a way for a user to know if the socket is under destruction? or is it enough to let the user track CLOSE/TERMINATED events?

Reply:
I would leave this to user to track timeline after `xlio_socket_destroy()` and before `XLIO_SOCKET_EVENT_TERMINATED`. Adding such information would unnecessary complicate the API. Better rely on the `xlio_socket_send()` failure, because inner check for `is_rts()` will fail.